tests/algorithms/test_required_on_load.py

```diff
@@ -97,3 +100,63 @@ def test_idempotent(algo_name: str):
         ).state.model
         compare_models(original_model, applied_twice_model, is_equal=False)  # Surgery actually changes model
         compare_models(applied_once_model, applied_twice_model, is_equal=True)  # Multiple applications are no-ops
+
+
+@pytest.mark.parametrize('algo_name', algorithms.__all__)
+@pytest.mark.parametrize('load_weights_only,already_added,exclude', [
+    [False, False, False],
+    [True, False, False],
+    [False, True, False],
+    [False, False, True],
+])
+def test_autoload(algo_name: str, load_weights_only: bool, already_added: bool, exclude: bool, tmp_path: pathlib.Path):
+    algo_cls = getattr(algorithms, algo_name)
+    if issubclass(algo_cls, Algorithm) and algo_cls.required_on_load():
+        algorithm = initialize_algorithm(algo_cls)
+
+        original_model = None
+        if algo_name == 'StochasticDepth':
+            original_model = composer_resnet(model_name='resnet50')
+        elif algo_name in ['Alibi', 'GatedLinearUnits']:
+            pytest.importorskip('transformers')
+            original_model = create_bert_classification()
+        else:
+            original_model = ConvModel()
+
+        trainer1 = Trainer(model=copy.deepcopy(original_model),
+                           algorithms=algorithm,
+                           save_folder=str(tmp_path),
+                           save_filename='ckpt.pt')
+        checkpoint_saver = [cb for cb in trainer1.state.callbacks if isinstance(cb, CheckpointSaver)][0]
+        checkpoint_saver._save_checkpoint(trainer1.state, trainer1.logger)
+
+        context = contextlib.nullcontext()
+        # Emit warning when autoloading
+        if not already_added and not exclude:
+            context = pytest.warns(UserWarning, match='Automatically adding required_on_load algorithm*')
+        # Excluding some algorithms leads to errors when loading
+        elif exclude:
+            if algo_name in ['Factorize', 'SqueezeExcite']:
```

Comment:
Is it possible to get nicer errors for these two cases? or its not possible to know something is wrong until you hit the error?

Reply:
Not really, you hit some weird stuff like optimizer mismatches params bc the surgery is missing

Reply:
If you explicitly exclude the algorithm and then hit the error, I feel like thats kinda on you

Reply:
yeah, thats fair :+1: